dhall-lsp-server/src/Dhall/LSP/Handlers.hs

```diff
@@ -2,59 +2,346 @@ module Dhall.LSP.Handlers where
+import qualified Network.URI as URI
+import qualified Network.URI.Encode as URI
+import Text.Megaparsec (SourcePos(..), unPos)
 
--- handler that doesn't do anything. Useful for example to make haskell-lsp shut
--- up about unhandled DidChangeTextDocument notifications (which are already
--- handled haskell-lsp itself).
-nullHandler :: LSP.LspFuncs () -> a -> IO ()
-nullHandler _ _ = return ()
 
-{- Currently implemented by the dummy nullHandler:
-initializedHandler :: LSP.LspFuncs () -> J.InitializedNotification -> IO ()
+-- Workaround to make our single-threaded LSP fit dhall-lsp's API, which
+-- expects a multi-threaded implementation. Reports errors to the user via the
+-- LSP `ShowMessage` notification.
+wrapHandler
+  :: MVar ServerState
+  -> (a -> HandlerM ())
+  -> a
+  -> IO ()
+wrapHandler vstate handle message =
+  modifyMVar_ vstate $
+    execStateT . runExceptT $
+      catchE (handle message) lspUserMessage
+
+lspUserMessage :: (Severity, Text) -> HandlerM ()
+lspUserMessage (Log, text) =
+  lspSendNotification LSP.NotLogMessage J.WindowLogMessage
+    $ J.LogMessageParams J.MtLog text
+lspUserMessage (severity, text) =
+  lspSendNotification LSP.NotShowMessage J.WindowShowMessage
+    $ J.ShowMessageParams severity' text
+  where severity' = case severity of
+          Error -> J.MtError
+          Warning -> J.MtWarning
+          Info -> J.MtInfo
+          Log -> J.MtLog
+
+
+lspSend :: LSP.FromServerMessage -> HandlerM ()
+lspSend msg = do
+  send <- lift $ gets (view (lspFuncs . sendFunc))
+  liftIO $ send msg
+
+lspRespond :: (J.ResponseMessage response -> LSP.FromServerMessage)
+  -> J.RequestMessage J.ClientMethod request response -> response -> HandlerM ()
+lspRespond constructor request response =
+  lspSend . constructor $ LSP.makeResponseMessage request response
+
+lspSendNotification
+  :: (J.NotificationMessage J.ServerMethod params -> LSP.FromServerMessage)
+  -> J.ServerMethod -> params -> HandlerM ()
+lspSendNotification constructor method params =
+  lspSend . constructor $ J.NotificationMessage "2.0" method params
+
+lspRequest
+  :: (J.RequestMessage J.ServerMethod params response -> LSP.FromServerMessage)
+  -> J.ServerMethod -> params -> HandlerM ()
+lspRequest constructor method params = do
+  getNextReqId <- lift $ gets (LSP.getNextReqId . view lspFuncs)
+  reqId <- liftIO getNextReqId
+  lspSend . constructor $ J.RequestMessage "2.0" reqId method params
+
+-- | A helper function to query haskell-lsp's VFS.
+readUri :: J.Uri -> HandlerM Text
+readUri uri = do
+  getVirtualFileFunc <- lift $ gets (LSP.getVirtualFileFunc . view lspFuncs)
+  asd <- liftIO $ getVirtualFileFunc uri
+  case asd of
+    Just (LSP.VirtualFile _ rope) -> return (Rope.toText rope)
+    Nothing -> fail $ "Could not find " <> show uri <> " in VFS."
+
+loadFile :: J.Uri -> HandlerM (Expr Src X)
+loadFile uri = do
+  txt <- readUri uri
+  fileIdentifier <- fileIdentifierFromUri uri
+  cache <- lift $ gets (view importCache)
```

Comment:
There is a `use` utility which you can use:

```haskel
  cache <- lift (use importCache)
```

Similarly, for something like:

```haskell
getVirtualFileFunc <- lift $ gets (LSP.getVirtualFileFunc . view lspFuncs)
```

... you can do:

```haskell
getVirtualFileFunc <- lift $ use (lspFuncs . to LSP.getVirtualFileFunc)
```

... or:

```haskell
getVirtualFileFunc <- lift $ uses lspFuncs LSP.getVirtualFileFunc
```

In fact, if you use `lens`/`lens-family` instead of `lens-family-core` then `use` is generalized to work with anything that implements `MonadState`, so then you don't even need the `lift`:

http://hackage.haskell.org/package/lens-family-1.2.3/docs/Lens-Family2-State-Lazy.html#v:use

Reply:
Since `dhall-lsp-server` is already depending on `lens`: `cache <- use importCache`. Much nicer!